src/gossip/service.go

```diff
@@ -125,7 +126,7 @@ func (s *Service) processEvent(realEngine Consensus, e *inter.Event) error {
 	// s.engineMu is locked here
 
 	if s.store.HasEvent(e.Hash()) { // sanity check
-		s.store.Log.Crit("Event is already processed", "event", e.Hash().String())
+		return event_check.ErrAlreadyConnectedEvent
```

Comment:
double-check signature and parents of block e. Avoid possibly a dup hash.

Reply:
signature check is very expensive, it may be checked only once

Reply:
OK. Make sure if it was checked. Where is the check?